### PR TITLE
BUG: arch/ABI fixes, code coverage improvements, and GH Actions tweaks

### PR DESCRIPTION
**Pull request by pcmoore**
This is a good example of a PR that started innocently enough but quickly reached the "whoa" stage, or as the folks like to say these days: that escalated quickly.

Originally this PR was supposed to be some minor cleanup and testing tweaks to try and bolster our code coverage numbers, however that uncovered a problem where the MIPS, SH, and PPC arch/ABIs were missing the `socketcall(2)` and `ipc(2)` multiplexed syscall handling.  In the process of fixing that I realized we were growing a *lot* of duplicated code relating to the multiplexed syscalls so I attempted to resolve that.  However in the process of consolidating the syscall multiplexing code I ran afoul of some "magic numbers" in that code and decided it would be best to get rid of that ... and so it went.

I'm marking this as for the v2.5.2 milestone simply because of the bug fixes contained in this PR, however this PR will obviously not apply directly to the release-2.5 branch.  Once this PR is past review and merged into the main branch, I'll backport this (and likely the necessary GH Actions changes) to the release-2.5 branch and submit a new PR.

### CODE REVIEW

File: tests/util.c [hunk 1]
```diff
@@ -47,6 +47,46 @@ static void _trap_handler(int signal, siginfo_t *info, void *ctx)
...
+ * @param ctx the filter context
+ * @param action the action for the rules
+ *
+ * This function is to make it easier for developers to temporarily add support
+ * for gcov/lcov to a test program; it likely should not be used in the normal
+ * regression tests.  Further, this should only be necessary for the "live"
+ * tests.
+ *
+ */
+int util_gcov_rules(const scmp_filter_ctx ctx, int action)
```
Comment (drakenclimber) [inline tests/util.c:61]: nitpick - `action` isn't needed.

**Review by drakenclimber — COMMENTED** [review]
The changes look good to me, @pcmoore.  I had a couple questions and nitpicky comments, but I think these changes are definitely an improvement.  And thanks for clearing up the code coverage quirks; that was bugging me.

File: src/arch-mips.c [hunk 2]
```diff
@@ -62,21 +99,466 @@ int mips_syscall_resolve_name_munge(const char *name)
...
+		return true;
+
+	name = mips_syscall_resolve_num(sys);
+	if (!name)
+		return false;
+
+#define _ABI_SYSCALL_IPC_CHK(NAME) \
+	if (!strcmp(name, #NAME)) return true;
+
+	_ABI_SYSCALL_IPC_CHK(semop)
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/arch-mips.c]: Check 'src/syscalls.c' in commit 0a789fa903f440b2c1effffb68f4660538e4f08a.  That commit consolidates most of the arch/ABI syscall multiplex code into a single implementation.  The only code that remains in 'src/arch-mips.c' are some cpp defines and the `struct arch_def` definitions for the ABIs.

[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/arch-mips.c]: Thanks.  I looked at that patch but totally missed the significance of it

File: src/arch-mips.c [hunk 3]
```diff
@@ -22,14 +22,21 @@
```
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/arch-mips.c:22]: > I looked through the entire file. I like the small cleanups you did compared to other multiplexed architectures.
> 
> A couple questions:
> 
> 1. Do we have a one-off way to test this (and other arch changes)?  (I can't recall which architectures you have access to.)  Anyway, it looks good to me... but so did the previous revision :/

The only arch/ABIs I have access to are the x86 variants and aarch64 :/

> 2. Any ideas on how/if we can reduce the large code duplication in the `rule_add()` functions across the various architectures?

Look at commit https://github.com/seccomp/libseccomp/commit/0a789fa903f440b2c1effffb68f4660538e4f08a and the `src/syscalls.c:abi_rule_add()` function :)

File: tests/util.c (hunk 1 above)
Comment (pcmoore) [reply tests/util.c:61]: Whops, it should be, nice catch :)  I'll fix that up after you've had a chance to look at the other comments I made (see above); if I force push now I'll likely blow away those commit links ...

File: src/arch-mips.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (drakenclimber) [reply src/arch-mips.c:22]: Commit https://github.com/seccomp/libseccomp/commit/0a789fa903f440b2c1effffb68f4660538e4f08a definitely resolved my concerns.  Thanks.  (I had written that comment before I had made it to commit https://github.com/seccomp/libseccomp/commit/0a789fa903f440b2c1effffb68f4660538e4f08a in the patchset.)

File: tests/util.c (hunk 1 above)
Comment (drakenclimber) [reply tests/util.c:61]: You're good to go on the force push

File: src/arch-mips.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (pcmoore) [reply src/arch-mips.c:22]: Cool, it's good to know we had the same thought and you're happy with the resolution.  I'm going to go ahead and mark this comment as resolved; the difficulty is supporting/test arches/ABIs is real, but unfortunately not new.

File: tests/util.c (hunk 1 above)
Comment (pcmoore) [reply tests/util.c:61]: Thanks.

This *should* be fixed now, but if you notice anything else awry let me know.

**Comment by drakenclimber** [comment]
Looks good to me.  Thanks, @pcmoore.
`Acked-by: Tom Hromatka <tom.hromatka@oracle.com>`

One random comment - I went to to the [coveralls results](https://coveralls.io/jobs/84804823) for your build, and it wouldn't expand the `src/` tree.  I've seen this before, so I don't think it's related to this change, but it may be worth watching.  The coverage data did look good though.

**Review by drakenclimber — APPROVED** [review]

**Comment by pcmoore** [comment]
> Looks good to me. Thanks, @pcmoore.
> `Acked-by: Tom Hromatka <tom.hromatka@oracle.com>`

Thanks for the review, I'll add your ACK, merge the patches, and start on a PR backport to release-2.5.

> One random comment - I went to to the [coveralls results](https://coveralls.io/jobs/84804823) for your build, and it wouldn't expand the `src/` tree. I've seen this before, so I don't think it's related to this change, but it may be worth watching. The coverage data did look good though.

I did notice that if I clicked on the "Back to job ..." it works.  Odd.

I've definitely seem some strange things with Coveralls and GH Actions, that is one of the reasons why I bundle the lcov HTML output in the code coverage artifacts now.

**Comment by pcmoore** [comment]
Merged at HEAD fcc601279004a7f4c2f6ebf766acb4556b0f5e65.

**Pull request closed without merging**